x-pack/plugins/osquery/server/routes/usage/recorder.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SavedObjectsClientContract } from 'kibana/server';
+import { usageMetricSavedObjectType } from '../../../common/types';
+import { LiveQuerySessionUsage } from '../../usage/types';
+
+export interface RouteUsageMetric {
+  queries: number;
+  errors: number;
+}
+
+export type RouteString = 'live_query';
+
+export const routeStrings: RouteString[] = ['live_query'];
+
+export async function createMetricObjects(soClient: SavedObjectsClientContract) {
+  const res = await Promise.allSettled(
+    routeStrings.map(async (route) => {
+      try {
+        await soClient.get(usageMetricSavedObjectType, route);
+      } catch (e) {
+        await soClient.create(
+          usageMetricSavedObjectType,
+          {
+            errors: 0,
+            count: 0,
+          },
+          {
+            id: route,
+          }
+        );
+      }
+    })
+  );
+  return !res.some((e) => e.status === 'rejected');
+}
+
+export async function getCount(soClient: SavedObjectsClientContract, route: RouteString) {
+  return await soClient.get<LiveQuerySessionUsage>(usageMetricSavedObjectType, route);
+}
+
+export interface CounterValue {
+  count: number;
+  errors: number;
+}
+
+export async function incrementCount(
+  soClient: SavedObjectsClientContract,
+  route: RouteString,
+  key: keyof CounterValue = 'count',
+  increment = 1
+) {
+  const metric = await soClient.get<CounterValue>(usageMetricSavedObjectType, route);
+  metric.attributes[key] += increment;
+  await soClient.update(usageMetricSavedObjectType, route, metric.attributes);
```

Comment:
Does this ever flush the state, or does it just increment forever?

Reply:
this is a cumulative count. i'm not entirely certain about the life times of saved objects, what events do they get flushed in outside of clearing them manually? i might be able to do something like clearing the count on fetch, but it felt like a wash to me. 

are there any situations where delta metrics would be better for this kind of thing?

Reply:
Thanks for bearing with me. I have been OOO for a couple of days :-)

> what events do they get flushed in outside of clearing them manually? 

They don't - unless a space is deleted AFAIK. 

> are there any situations where delta metrics would be better for this kind of thing?

Delta is maybe not a good word to describe it, but from what I have seen the usage collector runs every 24 hr and developers typically try to get a snapshot of the last 24 hours. That makes me think that clearing the count on read is what is needed. But I guess it only matters to people / systems consuming the telemetry. It's your call :- )

Reply:
thanks for the context! i feel better about changing it knowing there's prior art, and knowing about the collection interval.